lib/teleterm/daemon/daemon.go

```diff
@@ -1074,6 +1076,28 @@ func (s *Service) UpdateUserPreferences(ctx context.Context, clusterURI uri.Reso
 	return preferences, trace.Wrap(err)
 }
 
+// AuthenticateWebDevice is used to upgrade a web session (identified by a DeviceWebToken) to include device trust extensions.
+func (s *Service) AuthenticateWebDevice(ctx context.Context, rootClusterURI uri.ResourceURI, req *api.AuthenticateWebDeviceRequest) (*api.AuthenticateWebDeviceResponse, error) {
+	proxyClient, err := s.GetCachedClient(ctx, rootClusterURI)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	devicesClient := proxyClient.CurrentCluster().DevicesClient()
+
+	ceremony := dtauthn.NewCeremony()
```

Comment:
> Could we do a happy-path test? Doesn't have to be in-depth, for example we could fake the ceremony and just take a look at the input.

How do we fake the ceremony here though? Do we introduce a new interface implemented by two structs, one that does `dtauthn.NewCeremony()` and another which returns a fake ceremony in tests?

Even once that is solved, I'm not sure how we address `proxyClient.CurrentCluster().DevicesClient()`. Currently, some tests that don't actually need to execute stuff from clients are written to use `&client.ProxyClient{}` as `proxyClient`. I'm not sure if we'll be able to just call `CurrentCluster().DevicesClient()` on that.

https://github.com/gravitational/teleport/blob/2df6f6f8ccb3da16c64b9976828a410e2e480f83/lib/teleterm/daemon/daemon_test.go#L270-L278

https://github.com/gravitational/teleport/blob/2df6f6f8ccb3da16c64b9976828a410e2e480f83/lib/teleterm/daemon/daemon_test.go#L747-L749

Reply:
Yes, that would take a bit more effort to make sure it all works. We could do `s.newDeviceAuthnCeremony()` (so we can swap that function) and we would need `proxyClient.CurrentCluster().DevicesClient()` to at least not panic.

The ceremony interface can be defined here, it only needs the RunWeb() method.

Alternatively we could use the fakes at [lib/devicetrust/testenv][1]. In that case we just need to make sure the DevicesClient() used is the one it provides. Test setup is pretty much the same as [TestCeremony_RunWeb][2].

No strong opinions on my part, whatever y'all think is best works for me here.

[1]: https://github.com/gravitational/teleport/blob/4713707b2d0b4541d5a6f21dc37c3b3128896dc9/lib/devicetrust/testenv/testenv.go#L69

[2]: https://github.com/gravitational/teleport/blob/4713707b2d0b4541d5a6f21dc37c3b3128896dc9/lib/devicetrust/authn/authn_test.go#L103